src/utils/tcti.rs

```diff
@@ -0,0 +1,451 @@
+const DEVICE: &str = "device";
+const MSSIM: &str = "mssim";
+const TABRMD: &str = "tabrmd";
+
+// Possible TCTI to use with the ESYS API.
+/// Placeholder TCTI types that can be used when initialising a `Context` to determine which
+/// interface will be used to communicate with the TPM.
+#[derive(Clone, Debug, PartialEq)]
+pub enum Tcti {
+    /// Connect to a TPM available as a device node on the system
+    ///
+    /// For more information about configuration, see [this page](https://www.mankier.com/3/Tss2_Tcti_Device_Init)
+    Device(DeviceConfig),
+    /// Connect to a TPM (simulator) available as a network device
+    ///
+    /// For more information about configuration, see [this page](https://www.mankier.com/3/Tss2_Tcti_Mssim_Init)
+    Mssim(MssimConfig),
+    /// Connect to a TPM through an Access Broker/Resource Manager daemon
+    ///
+    /// For more information about configuration, see [this page](https://www.mankier.com/3/Tss2_Tcti_Tabrmd_Init)
+    Tabrmd(TabrmdConfig),
+}
+
+impl TryFrom<Tcti> for CString {
+    type Error = Error;
+
+    fn try_from(tcti: Tcti) -> Result<Self, Error> {
+        let tcti_name = match tcti {
+            Tcti::Device(..) => DEVICE,
+            Tcti::Mssim(..) => MSSIM,
+            Tcti::Tabrmd(..) => TABRMD,
+        };
+
+        let tcti_conf = match tcti {
+            Tcti::Mssim(config) => {
+                if let Some(ServerAddress::Hostname(name)) = &config.host {
+                    if !hostname_validator::is_valid(name) {
+                        return Err(Error::WrapperError(WrapperErrorKind::InvalidParam));
+                    }
+                }
+                format!(
+                    ":host={},port={}",
+                    config.host.unwrap_or_default(),
+                    config.port.unwrap_or(2321)
+                )
+            }
+            Tcti::Device(DeviceConfig { path: Some(path) }) => format!(
+                ":{}",
+                path.to_str()
+                    .ok_or(Error::WrapperError(WrapperErrorKind::InvalidParam))?,
+            ),
+            Tcti::Device(DeviceConfig { path: None }) => String::default(),
+            Tcti::Tabrmd(config) => format!(
+                ":bus_name={},bus_type={}",
+                config
+                    .bus_name
+                    .unwrap_or_else(|| String::from("com.intel.tss2.Tabrmd")),
+                config.bus_type.unwrap_or(BusType::System)
+            ),
+        };
+
+        CString::new(format!("{}{}", tcti_name, tcti_conf))
+            .or(Err(Error::WrapperError(WrapperErrorKind::InvalidParam)))
+    }
+}
+
+impl FromStr for Tcti {
+    type Err = Error;
+
+    fn from_str(config_str: &str) -> Result<Self, Self::Err> {
+        let device_pattern = Regex::new(r"^device(:(.*))?$").unwrap(); //should not fail
+        if device_pattern.is_match(config_str) {
+            match device_pattern.captures(config_str) {
+                None => unreachable!("Branch can never be reached"),
```

Comment:
Why is it unreachable here? What if `config_str` is only `"device"`?

Reply:
It's unreachable because I check two lines above that there is a match - even if `config_str` is only "device", there is at least one capture group, which is the whole matched string (in this case "device").

Reply:
Would it be some kind of performance hit to do:
```rust
if let Some(captures) = device_pattern.captures(config_str) {
    return Ok(Tcti::Device(DeviceConfig::from_str(
                     captures.get(2).map_or("", |m| m.as_str()),
                     )?));
}
```
Because I think that would look nicer.

Reply:
That sounds good!